### DISCUSSION TITLE
Python: Strange Recursion Error Occurs in simple code

### DISCUSSION THREAD

**Discussion by elliteDanger** [community/community#138356] [Programming Help] [answered]
### Body

Hello, guys! Hope you are doing well.
I just wrote a simple python code snippet as part of bigger project.
I would like to deal with file system, specifically directories.
The code is much more complex but I am just simplifying it for convenience.

class Directory:
    def __init__(self, name):
        self.name = name
        self.subdirectories = []

    def add_subdirectory(self, subdirectory):
        self.subdirectories.append(subdirectory)

def print_directory_structure(directory, level=0):
    print(" " * level * 2 + directory.name)
    for subdirectory in directory.subdirectories:
        print_directory_structure(subdirectory, level + 1)

# Create a deep directory structure
root = Directory("root")
current = root
for i in range(1000):  # Creating a deep structure
    new_dir = Directory(f"dir_{i}")
    current.add_subdirectory(new_dir)
    current = new_dir

print_directory_structure(root)

So in essence, I am just creating complex sub-nested directory (to simulate the real situation of the current file system), and tried to output the tree system.
But I get this error.

Traceback (most recent call last):
  File "directory_structure.py", line 14, in <module>
    print_directory_structure(root)
  File "directory_structure.py", line 7, in print_directory_structure
    print_directory_structure(subdirectory, level + 1)
  ...
RecursionError: maximum recursion depth exceeded in comparison

I encounter this type of error for the first time, although I don't have that much experience with Python.
I can deal with common errors but this one is new to me so I want to learn from professional.
This would be pretty much straightforward for a senior.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by razeenf** [comment]
The error you're encountering, occurs because Python has a default recursion limit. When your recursive function goes too deep (i.e., too many levels of nested subdirectories), Python hits this limit and raises the error.

In your case, since you're creating a directory structure with 1000 levels of nesting, your recursive function exceeds the default recursion depth. To solve this you can either increase the recursive limit for python using sys.setrecursionlimit or switch to an iterative approach instead of using recursion.

**Comment by tpmst** [comment]
To resolve this issue, you have a couple of options:

### 1. Increase the Recursion Limit
You can increase the maximum recursion limit using sys.setrecursionlimit(). However, this is generally not recommended unless absolutely necessary because it can lead to crashes or stack overflow errors if the recursion goes too deep.

Here's how you would increase the recursion limit:

```
import sys

sys.setrecursionlimit(1500)  # Setting the recursion limit to 1500

class Directory:
    def __init__(self, name):
        self.name = name
        self.subdirectories = []

    def add_subdirectory(self, subdirectory):
        self.subdirectories.append(subdirectory)

def print_directory_structure(directory, level=0):
    print(" " * level * 2 + directory.name)
    for subdirectory in directory.subdirectories:
        print_directory_structure(subdirectory, level + 1)

# Create a deep directory structure
root = Directory("root")
current = root
for i in range(1000):
    new_dir = Directory(f"dir_{i}")
    current.add_subdirectory(new_dir)
    current = new_dir

print_directory_structure(root)
```



### 2. Refactor to Use Iteration Instead of Recursion
A more robust solution would be to refactor your print_directory_structure function to use an iterative approach rather than a recursive one. This can be done using a stack or a queue, depending on how you want to traverse the structure (depth-first or breadth-first). Here's an example using a depth-first search (DFS) with a stack:

```
class Directory:
    def __init__(self, name):
        self.name = name
        self.subdirectories = []

    def add_subdirectory(self, subdirectory):
        self.subdirectories.append(subdirectory)

def print_directory_structure(directory):
    stack = [(directory, 0)]  # Stack to hold directory and current level
    while stack:
        current_directory, level = stack.pop()
        print(" " * level * 2 + current_directory.name)
        for subdirectory in reversed(current_directory.subdirectories):
            stack.append((subdirectory, level + 1))

# Create a deep directory structure
root = Directory("root")
current = root
for i in range(1000):
    new_dir = Directory(f"dir_{i}")
    current.add_subdirectory(new_dir)
    current = new_dir

print_directory_structure(root)
```

## Conclusion

Using an iterative approach to handle deep directory structures avoids the risk of exceeding the recursion limit and is generally more robust for handling large data structures in Python. Adjusting the recursion limit might be a quick fix, but it's usually better to refactor the code to avoid recursion where possible, especially when dealing with potentially very deep or large structures.

**Comment by elliteDanger** [comment]
I personally don't want to increate the limit itself. I don't want to break or change what it was meant to be.
Can someone guide simpler solution for me? It's pretty much confusing for me..

**Comment by leoyoung02** [comment, marked as answer]
Reason: RecursionError occurs because the function print_directory_structure exceeds the maximum recursion depth set by Python.



`stack = [(root, 0)]  # Stack to hold directories and their level`

`current_directory, level = stack.pop()`

`for subdirectory in reversed(current_directory.subdirectories):`



Those are key lines to consider in print_directory_structure_iterative function.
If you use stack mechanism, you can avoid the recursion error.


**Important**: Focus on part using reversed(current_directory.subdirectoeis). The subdirectories are pushed onto the stack in reverse order to maintain the correct order of processing when popped off the stack. Remember we used stack to deal with this problem.


Feel free to ask if you have any more concerns or confusing parts.

**Comment by elliteDanger** [comment]
Wow. That looks much simpler. So using stack and iterate all the subdirectories, step by step right?
That looks much like BFS or something.

**Comment by MatheusDiamantino** [comment]
The error you're encountering, RecursionError: maximum recursion depth exceeded, happens because Python has a default recursion limit. When a function calls itself too many times without reaching a base case (in this case, iterating through a deeply nested directory structure), it hits this limit and throws an error.

You're creating a deeply nested structure with 1,000 directories, which causes the recursive function print_directory_structure to exceed the recursion limit.

Solution Options:
Increase the recursion limit: You can increase Python’s recursion limit using the sys module. However, this is not always recommended because if your structure becomes even deeper, you might still hit the limit. Here's how to increase the recursion limit:

python
Copiar código
import sys
sys.setrecursionlimit(2000)  # Set a higher limit
Place this at the top of your script. This will allow the recursive function to handle deeper structures, but there is still a risk of reaching the new limit with more complex structures.

Switch to an iterative approach: The better solution would be to refactor your code to avoid recursion and instead use an iterative method, which won't be limited by recursion depth.

Here’s how you can modify your print_directory_structure to use a loop instead of recursion:

python
Copiar código
def print_directory_structure(directory):
    stack = [(directory, 0)]  # Stack to hold (directory, level)
    while stack:
        current_dir, level = stack.pop()
        print(" " * level * 2 + current_dir.name)
        for subdirectory in reversed(current_dir.subdirectories):
            stack.append((subdirectory, level + 1))
This uses a stack to simulate the recursive call stack but without the limitations of recursion depth. It will print the structure as expected without hitting a recursion error, even for deeply nested directories.

Why this works:
By using a stack (LIFO structure), we avoid recursion and handle deeply nested directories in a more controlled manner.
This iterative approach is much more scalable for complex structures.
Summary:
Option 1: Increase recursion limit temporarily using sys.setrecursionlimit().
Option 2 (recommended): Refactor the code to use an iterative method with a stack.
The iterative method is more robust and avoids hitting Python's recursion limits altogether, making it the preferred approach in this case.

**Comment by MatheusDiamantino** [comment]
Response #10108

**Comment by leoyoung02** [comment]
Yep. That's right. I think you got the right point. This approach will resolve the error.

**Comment by elliteDanger** [comment]
Wow. Thanks a lot, all of you. It's working fine now.
